src/vs/workbench/api/electron-browser/mainThreadStorage.ts

```diff
@@ -3,47 +3,74 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
-import { TPromise } from 'vs/base/common/winjs.base';
 import { IStorageService, StorageScope } from 'vs/platform/storage/common/storage';
-import { MainThreadStorageShape, MainContext, IExtHostContext } from '../node/extHost.protocol';
+import { MainThreadStorageShape, MainContext, IExtHostContext, ExtHostStorageShape, ExtHostContext } from '../node/extHost.protocol';
 import { extHostNamedCustomer } from 'vs/workbench/api/electron-browser/extHostCustomers';
+import { IDisposable } from 'vs/base/common/lifecycle';
 
 @extHostNamedCustomer(MainContext.MainThreadStorage)
 export class MainThreadStorage implements MainThreadStorageShape {
 
 	private _storageService: IStorageService;
+	private _proxy: ExtHostStorageShape;
+	private _storageListener: IDisposable;
+	private _sharedStorageKeysToWatch: Map<string, boolean> = new Map<string, boolean>();
 
 	constructor(
 		extHostContext: IExtHostContext,
 		@IStorageService storageService: IStorageService
 	) {
 		this._storageService = storageService;
+		this._proxy = extHostContext.getProxy(ExtHostContext.ExtHostStorage);
+
+		this._storageListener = this._storageService.onDidChangeStorage(e => {
+			let shared = e.scope === StorageScope.GLOBAL;
+			if (shared && this._sharedStorageKeysToWatch.has(e.key)) {
+				let value = this._getValue(shared, e.key);
+				if (value instanceof Error) {
+					return; // ignore parsing errors that can happen
+				}
+				this._proxy.$acceptValue(shared, e.key, value);
+			}
+		});
 	}
 
 	dispose(): void {
+		this._storageListener.dispose();
 	}
 
 	$getValue<T>(shared: boolean, key: string): Thenable<T> {
+		if (shared) {
+			this._sharedStorageKeysToWatch.set(key, true);
```

Comment:
We can have this dynamic set but effectively this is _every_ installed extension. So instead of this we can ask the extension service if a key is an extension id.

Reply:
for reference: https://github.com/Microsoft/vscode/blob/7688411a9c916fdc42071e2dc13145796437da83/src/vs/workbench/api/node/extHostExtensionService.ts#L39

Reply:
Yeah, that's what happens at runtime. *Before* the extension host loads extensions it has read all their mementos and keeps them so that they can be read sync. 

So, whatever you prefer. Having a map looks like a nice encapsulation but effectively it will have all extension ids soon after the start and it will never change after that again.

Reply:
@jrieken yeah I chose the map so that mainThreadStorage does not need to be aware of the fact that only extension IDs are used as keys. I thought it was weird to have that knowledge there given that file has a very generic API (key, value).